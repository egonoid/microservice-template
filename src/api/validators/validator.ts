import { ObjectSchema, ValidateOptions, ValidationError } from 'yup';

export const validate = async <T>(
  item: string,
  schema: ObjectSchema,
  options?: ValidateOptions
): Promise<T> => {
  if (!options) {
    options = {
      abortEarly: false,
      stripUnknown: true,
    };
  }

  const result = schema
    .validate(item, options)
    .then(value => {
      return value as Promise<T>;
    })
    .catch((err: ValidationError) => {
      if (err && err.inner) {
        const errors: any[] = [];
        err.inner.forEach(error => {
          const source = error.path;
          const type = getErrorType(error.value, error.type);
          const message = getErrorMessage(source, type, error.params);

          const newError = {
            source,
            type,
            message,
          };

          errors.push(newError);
        });

        throw { errors };
      }
      throw err;
    });

  return result;
};

const getErrorType = (value: string | number, type: string) => {
  if (typeof value === 'undefined') {
    switch (type) {
      case 'required':
        return 'ERROR_VALIDATION_REQUIRED';
    }
  }

  if (typeof value === 'string') {
    switch (type) {
      case 'min':
        return 'ERROR_VALIDATION_MIN_STRING';
      case 'max':
        return 'ERROR_VALIDATION_MAX_STRING';
      case 'email':
        return 'ERROR_VALIDATION_EMAIL';
    }
  }

  if (typeof value === 'number') {
    switch (type) {
      case 'min':
        return 'ERROR_VALIDATION_MIN_NUMBER';
      case 'max':
        return 'ERROR_VALIDATION_MAX_NUMBER';
    }
  }

  return 'ERROR_VALIDATION_UNKNOWN';
};

// inspired by https://github.com/jquense/yup/blob/master/src/locale.js
const getErrorMessage = (source: string, type: string, params: any) => {
  source = source.charAt(0).toUpperCase() + source.slice(1);

  switch (type) {
    case 'ERROR_VALIDATION_REQUIRED':
      return `${source} is a required field.`;
    case 'ERROR_VALIDATION_MIN_STRING':
      return `${source} must be at least ${params.min} characters long.`;
    case 'ERROR_VALIDATION_MAX_STRING':
      return `${source} cannot exceed ${params.max} characters.`;
    case 'ERROR_VALIDATION_EMAIL':
      return `${source} is not a valid email address.`;
    case 'ERROR_VALIDATION_MIN_NUMBER':
      return `${source} must be greater than or equal to ${params.min}.`;
    case 'ERROR_VALIDATION_MAX_NUMBER':
      return `${source} must be less than or equal to ${params.max}.`;
    default:
      return `${source} is invalid.`;
  }
};
