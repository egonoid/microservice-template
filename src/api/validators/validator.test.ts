import * as yup from 'yup';
import { validate } from './validator';

const model = {
  company: 'Super Inc.',
  email: 'test@example.com',
  malicious: 'Im unwanted...',
  firstName: 'Marc',
  lastName: 'Miller',
  street: 'Millerstreet. 27',
  zipCode: '12345',
  city: 'Example City',
  country: 'Tryout Country',
  age: 18,
};

const unitTestSchema = yup.object({
  type: yup.mixed<'person' | 'company'>().oneOf(['person', 'company']),
  company: yup
    .string()
    .required()
    .min(3, 'Company has minimum length of 3 letters.')
    .max(100, 'Company has maximum length of 100 letters.'),
  firstName: yup
    .string()
    .required('Entering a first name is required.')
    .min(3, 'First name has minimum length of 3 letters.')
    .max(100, 'First name has maximum length of 100 letters.'),
  lastName: yup
    .string()
    .required('Entering a last name is required.')
    .min(3, 'Last name has minimum length of 3 letters.')
    .max(100, 'Last name has maximum length of 100 letters.'),
  email: yup
    .string()
    .required('Entering an email is required.')
    .email('Please enter a valid email address.'),
  street: yup.string().required('Entering a street is required.'),
  streetNumber: yup.string(),
  zipCode: yup.string().required('Entering a zip code is required.'),
  city: yup.string().required('Entering a city is required.'),
  country: yup.string().required('Entering a country is required.'),
  state: yup.string(),
  phone: yup.string(),
  age: yup
    .number()
    .required()
    .min(18)
    .max(99),
});

type UnitTestModel = yup.InferType<typeof unitTestSchema>;

describe('validate UnitTestModel', () => {
  it('should validate without error', async () => {
    // Arrange
    const item = {
      ...model,
    };

    // Act
    const result = await validate<UnitTestModel>(
      JSON.stringify(item),
      unitTestSchema
    );

    // Assert
    expect(result).toBeTruthy();
  });

  it('should throw exception on email address', async () => {
    // Arrange
    const item = {
      ...model,
      email: 'test@example.com!',
    };

    expect.assertions(1);

    try {
      // Act
      await validate<UnitTestModel>(JSON.stringify(item), unitTestSchema);
    } catch (err) {
      // Assert
      expect(err).toMatchSnapshot();
    }
  });

  it('should throw exception on too short firstname', async () => {
    // Arrange
    const item = {
      ...model,
      firstName: 'An',
    };

    expect.assertions(1);

    try {
      // Act
      await validate<UnitTestModel>(JSON.stringify(item), unitTestSchema);
    } catch (err) {
      // Assert
      expect(err).toMatchSnapshot();
    }
  });

  it('should throw exception on too long firstname', async () => {
    // Arrange
    const item = {
      ...model,
      firstName:
        '12345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901',
    };

    expect.assertions(1);

    try {
      // Act
      await validate<UnitTestModel>(JSON.stringify(item), unitTestSchema);
    } catch (err) {
      // Assert
      expect(err).toMatchSnapshot();
    }
  });

  it('should throw exception on number to low', async () => {
    // Arrange
    const item = {
      ...model,
      age: 17,
    };

    expect.assertions(1);

    try {
      // Act
      await validate<UnitTestModel>(JSON.stringify(item), unitTestSchema);
    } catch (err) {
      // Assert
      expect(err).toMatchSnapshot();
    }
  });
});
