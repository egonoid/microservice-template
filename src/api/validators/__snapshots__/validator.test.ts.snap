// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validate UnitTestModel should throw exception on email address 1`] = `
Object {
  "errors": Array [
    Object {
      "message": "Email is not a valid email address.",
      "source": "email",
      "type": "ERROR_VALIDATION_EMAIL",
    },
  ],
}
`;

exports[`validate UnitTestModel should throw exception on number to low 1`] = `
Object {
  "errors": Array [
    Object {
      "message": "Age must be greater than or equal to 18.",
      "source": "age",
      "type": "ERROR_VALIDATION_MIN_NUMBER",
    },
  ],
}
`;

exports[`validate UnitTestModel should throw exception on too long firstname 1`] = `
Object {
  "errors": Array [
    Object {
      "message": "FirstName cannot exceed 100 characters.",
      "source": "firstName",
      "type": "ERROR_VALIDATION_MAX_STRING",
    },
  ],
}
`;

exports[`validate UnitTestModel should throw exception on too short firstname 1`] = `
Object {
  "errors": Array [
    Object {
      "message": "FirstName must be at least 3 characters long.",
      "source": "firstName",
      "type": "ERROR_VALIDATION_MIN_STRING",
    },
  ],
}
`;
